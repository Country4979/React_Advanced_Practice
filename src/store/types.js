export const AUTH_LOGIN = 'AUTH_LOGIN';
export const AUTH_LOGIN_REQUEST = 'AUTH_LOGIN_REQUEST';
export const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
export const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';
export const SET_TOKEN = 'SET_TOKEN';

export const UI_RESET_ERROR = 'UI_RESET_ERROR';

export const ADVERTS_LOADED_REQUEST = 'ADVERTS_LOADED_REQUEST';
export const ADVERTS_LOADED_SUCCESS = 'ADVERTS_LOADED_SUCCESS';
export const ADVERTS_LOADED_FAILURE = 'ADVERTS_LOADED_FAILURE';

export const ADVERT_LOADED_REQUEST = 'ADVERT_LOADED_REQUEST';
export const ADVERT_LOADED_SUCCESS = 'ADVERT_LOADED_SUCCESS';
export const ADVERT_LOADED_FAILURE = 'ADVERT_LOADED_FAILURE';

export const ADVERT_CREATED_SUCCESS = 'ADVERT_CREATED_SUCCESS';
export const ADVERT_CREATED_REQUEST = 'ADVERT_CREATED_REQUEST';
export const ADVERT_CREATED_FAILURE = 'ADVERT_CREATED_FAILURE';

export const ADVERT_DELETED_SUCCESS = 'ADVERT_DELETED_SUCCESS';
export const ADVERT_DELETED_REQUEST = 'ADVERT_DELETED_REQUEST';
export const ADVERT_DELETED_FAILURE = 'ADVERT_DELETED_FAILURE';

export const ADD_TAGS_REQUEST = 'ADD_TAGS_REQUEST';
export const ADD_TAGS_SUCCESS = 'ADD_TAGS_SUCCESS';
export const ADD_TAGS_FAILURE = 'ADD_TAGS_FAILURE';

export const ADV_FILTER_NAME = 'ADV_FILTER_NAME';
export const ADV_FILTER_SALE = 'ADV_FILTER_SALE';
export const ADV_FILTER_TAGS = 'ADV_FILTER_TAGS';
//export const ADV_FILTER_MIN_PRICE = 'ADV_FILTER_MIN_PRICE';
//export const ADV_FILTER_MAX_PRICE = 'ADV_FILTER_MAX_PRICE';
export const ADV_FILTER_PRICE = 'ADV_FILTER_PRICE';

export const TOGGLE_RESULT = 'TOGGLE_RESULT'; //Show filter results?
