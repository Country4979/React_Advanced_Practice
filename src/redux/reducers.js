export const defaultStore = {
    auth: false,
    adverts: [],
};
