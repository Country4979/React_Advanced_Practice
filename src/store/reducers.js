import {
    ADVERTS_LOADED_SUCCESS,
    ADVERT_LOADED_SUCCESS,
    ADVERT_CREATED_SUCCESS,
    ADVERT_DELETED_SUCCESS,
    AUTH_LOGIN_SUCCESS,
    SET_TOKEN,
    AUTH_LOGOUT,
    UI_RESET_ERROR,
    ADD_TAGS_SUCCESS,
    ADV_FILTER_NAME,
    //ADV_FILTER_SALE,
    //ADV_FILTER_TAGS,
    //ADV_FILTER_MIN_PRICE,
    //ADV_FILTER_MAX_PRICE,
    TOGGLE_RESULT,
    ADV_FILTER_PRICE,
    OPEN_MODAL,
    CLOSE_MODAL,
} from './types';

export const defaultState = {
    auth: false,
    token: null,
    adverts: {
        areLoaded: false,
        data: [],
    },
    tags: {
        areLoaded: false,
        data: [],
    },
    filtered: {
        query: '',
        querySale: '',
        queryTags: [],
        queryMinPrice: '',
        queryMaxPrice: '',
    },
    ui: {
        isLoading: false,
        error: null,
        dataFiltered: false,
    },
    modal: {
        isOpen: false,
        modalType: null,
    }
};

export function modal(state = defaultState.ui, action) {
    switch (action.type) {
        case OPEN_MODAL:
            return {
                ...state,
                isOpen: true,
                modalType: action.payload,
            };
        case CLOSE_MODAL:
            return defaultState;
        default:
            return state;
    }
}

export function auth(state = defaultState.auth, action) {
    switch (action.type) {
        case AUTH_LOGIN_SUCCESS:
            return true;
        case AUTH_LOGOUT:
            return false;
        default:
            return state;
    }
}
export function token(state = defaultState.token, action) {
    switch (action.type) {
        case SET_TOKEN:
            return action.payload;
        default:
            return state;
    }
}

export function adverts(state = defaultState.adverts, action) {
    if (action.type === ADVERTS_LOADED_SUCCESS) {
        return { areLoaded: true, data: action.payload };
    }
    if (action.type === ADVERT_LOADED_SUCCESS) {
        return { ...state, data: [action.payload] };
    }
    if (action.type === ADVERT_CREATED_SUCCESS) {
        return { ...state, data: [...state.data, action.payload] };
    }
    if (action.type === ADVERT_DELETED_SUCCESS) {
        // Look for the index of the advertisement we want to delete
        const index = state.data.findIndex(
            (advert) => advert.id === action.payload
        );
        // If we find the index, create a new array without the removed advertisement.
        if (index !== -1) {
            return {
                ...state,
                data: state.data.filter(
                    (advert) => advert.id !== action.payload
                ),
            };
        }
        // Create a copy of the original array and update the advertisement load in the state
        return {
            ...state,
            data: [...state.data],
        };
    }
    return state;
}

export function ui(state = defaultState.ui, action) {
    if (action.error) {
        return { isLoading: false, error: action.payload };
    }

    if (/_REQUEST$/.test(action.type)) {
        return { isLoading: true, error: null };
    }

    if (/_SUCCESS$/.test(action.type)) {
        return { isLoading: false, error: null };
    }

    if (action.type === UI_RESET_ERROR) {
        return { ...state, error: null };
    }
    if (action.type === TOGGLE_RESULT) {
        return {
            ...state,
            dataFiltered: action.value,
        };
    }
    return state;
}

export function tagsList(state = defaultState.tags, action) {
    if (action.type === ADD_TAGS_SUCCESS) {
        return { areLoaded: true, data: action.payload };
    }
    return state;
}
export function filteredAdverts(state = defaultState.filtered, action) {
    var filteredAdverts = [];
    switch (action.type) {
        /*case ADV_FILTER_NAME:
            const query = action.payload;
            if (state.adverts) {
                if (query === '') {
                    filteredAdverts = state.adverts;
                } else {
                    filteredAdverts = state.adverts.filter(ad => ad.name.toLowerCase().includes(query.toLowerCase()));
                }
            }
            return {
                ...state,
                query: filteredAdverts,
            };*/
        case ADV_FILTER_NAME:
        case ADV_FILTER_PRICE:
            const { query } = action.payload;
            const { minPrice, maxPrice } = action.payload;
            let filteredAdverts;
            if (state.adverts) {
                filteredAdverts = state.adverts;
                if (query !== undefined) {
                    filteredAdverts = filteredAdverts.filter((ad) =>
                        ad.name.toLowerCase().includes(query.toLowerCase())
                    );
                } else if (minPrice !== undefined && maxPrice !== undefined) {
                    filteredAdverts = filteredAdverts.filter(
                        (ad) => ad.price >= minPrice && ad.price <= maxPrice
                    );
                }
            }
            return {
                ...state,
                filteredAdverts,
            };

        /*case ADV_FILTER_SALE:
            return {
                ...state,
                querySale: action.payload,
            };
        case ADV_FILTER_TAGS:
            return {
                ...state,
                queryTags: action.payload,
            };
        case ADV_FILTER_MIN_PRICE:
            return {
                ...state,
                queryMinPrice: action.payload,
            };
        case ADV_FILTER_MAX_PRICE:
            return {
                ...state,
                queryMaxPrice: action.payload,
            };*/
        default:
            return state;
    }
}
