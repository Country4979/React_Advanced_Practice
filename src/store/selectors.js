export const getIsLogged = (state) => state.auth;
export const getToken = (state) => state.token;
export const areAdvertsLoaded = (state) => state.adverts.areLoaded;
export const getAdverts = (state) =>
    state.adverts.data ? state.adverts.data : [];

export const getAdvertById = (advertId) => (state) =>
    getAdverts(state).find((adverts) => adverts.id === advertId);

export const getAdvertsFilter = (state) => state.filtered;

export const getUi = (state) => state.ui;

export const getModalstate = (state) => state.modal;

export const getTags = (state) => state.tagsList.data;
export const areTagsLoaded = (state) => state.tagsList.areLoaded;
