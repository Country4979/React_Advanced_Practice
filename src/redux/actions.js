import {
    ADVERTS_LOADED_FAILURE,
    ADVERTS_LOADED_REQUEST,
    ADVERTS_LOADED_SUCCESS,
    AUTH_LOGIN_FAILURE,
    AUTH_LOGIN_REQUEST,
    AUTH_LOGIN_SUCCESS,
    ADVERT_LOADED_REQUEST,
    ADVERT_LOADED_SUCCESS,
    ADVERT_LOADED_FAILURE,
    AUTH_LOGOUT,
    UI_RESET_ERROR,
} from './types';
import { areAdvertsLoaded } from './selectors';

// LOGIN actions & thunk:
export const authLogin =
    (credentials) =>
    async (dispatch, _getState, { auth }) => {
        dispatch(authLoginRequest());
        try {
            await auth.login(credentials);
            //Logged in:
            dispatch(authLoginSuccess());
        } catch (error) {
            dispatch(authLoginFailure(error));
            throw error;
        }
    };

export const authLoginRequest = () => ({
    type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
    type: AUTH_LOGIN_SUCCESS,
});
export const authLoginFailure = (error) => ({
    type: AUTH_LOGIN_FAILURE,
    error: true,
    payload: error,
});

export const uiResetError = () => ({
    type: UI_RESET_ERROR,
});

//LOGOUT
export const authLogout = () => ({
    type: AUTH_LOGOUT,
});

// ADVERTS thunk & actions:

export const advertsLoaded =
    () =>
    async (dispatch, getState, { advs }) => {
        if (areAdvertsLoaded(getState())) {
            return;
        }

        dispatch(advertsLoadedRequest());
        try {
            const adverts = await advs.getLastAdv();
            dispatch(advertsLoadedSuccess(adverts));
        } catch (error) {
            dispatch(advertsLoadedFailure(error));
            throw error;
        }
    };

export const advertsLoadedRequest = () => ({
    type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = (adverts) => ({
    type: ADVERTS_LOADED_SUCCESS,
    payload: adverts,
});

export const advertsLoadedFailure = (error) => ({
    type: ADVERTS_LOADED_FAILURE,
    error: true,
    payload: error,
});

// ADVERT DETAIL thunk & actions:
export const advertLoaded =
    () =>
    async (dispatch, getState, { advs }) => {
        dispatch(advertsLoadedRequest());
        try {
            const adverts = await advs.getLastAdv();
            dispatch(advertsLoadedSuccess(adverts));
        } catch (error) {
            dispatch(advertsLoadedFailure(error));
            throw error;
        }
    };

export const advertLoadedRequest = () => ({
    type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = (adverts) => ({
    type: ADVERT_LOADED_SUCCESS,
    payload: adverts,
});

export const advertLoadedFailure = (error) => ({
    type: ADVERT_LOADED_FAILURE,
    error: true,
    payload: error,
});
