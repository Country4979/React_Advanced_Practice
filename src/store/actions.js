import {
    ADVERTS_LOADED_FAILURE,
    ADVERTS_LOADED_REQUEST,
    ADVERTS_LOADED_SUCCESS,
    AUTH_LOGIN_FAILURE,
    AUTH_LOGIN_REQUEST,
    AUTH_LOGIN_SUCCESS,
    SET_TOKEN,
    ADVERT_LOADED_REQUEST,
    ADVERT_LOADED_SUCCESS,
    ADVERT_LOADED_FAILURE,
    ADVERT_CREATED_REQUEST,
    ADVERT_CREATED_SUCCESS,
    ADVERT_CREATED_FAILURE,
    ADVERT_DELETED_REQUEST,
    ADVERT_DELETED_SUCCESS,
    ADVERT_DELETED_FAILURE,
    AUTH_LOGOUT,
    ADD_TAGS_REQUEST,
    ADD_TAGS_SUCCESS,
    ADD_TAGS_FAILURE,
    UI_RESET_ERROR,
    ADV_FILTER_NAME,
    ADV_FILTER_SALE,
    ADV_FILTER_TAGS,
    //ADV_FILTER_MIN_PRICE,
    //ADV_FILTER_MAX_PRICE,
    TOGGLE_RESULT,
    ADV_FILTER_PRICE,
    OPEN_MODAL,
    CLOSE_MODAL,
} from './types';
import { areAdvertsLoaded, areTagsLoaded, getAdvertById } from './selectors';

//TOKEN actions
export const setToken = (token) => ({
    type: SET_TOKEN,
    payload: token,
});

// LOGIN actions & thunk:
export const authLogin =
    (credentials) =>
    async (dispatch, _getState, { service: { auth }, router }) => {
        dispatch(authLoginRequest());
        try {
            await auth.login(credentials);

            //Logged in:
            dispatch(authLoginSuccess());

            // Set token in the store
            const token = localStorage.getItem('auth');
            dispatch(setToken(token));
            // Redirect to pathname
            const to = router.state?.from?.pathname || '/';
            router.navigate(to);
        } catch (error) {
            dispatch(authLoginFailure(error));
            throw error;
        }
    };

export const authLoginRequest = () => ({
    type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
    type: AUTH_LOGIN_SUCCESS,
});
export const authLoginFailure = (error) => ({
    type: AUTH_LOGIN_FAILURE,
    error: true,
    payload: error,
});

export const uiResetError = () => ({
    type: UI_RESET_ERROR,
});

//LOGOUT
export const authLogoutSuccess = () => ({
    type: AUTH_LOGOUT,
});

export const authlogout =
    () =>
    async (dispatch, _getState, { client, service: { auth }, router }) => {
        await auth.logout();
        client.removeAuthorizationHeader();
        dispatch(authLogoutSuccess());
        router.navigate('/login');
    };

// ADVERTS thunk & actions:

export const advertsLoaded =
    () =>
    async (dispatch, getState, { service: { advs } }) => {
        if (areAdvertsLoaded(getState())) {
            return;
        }
        dispatch(advertsLoadedRequest());
        try {
            const adverts = await advs.getLastAdv();
            dispatch(advertsLoadedSuccess(adverts));
        } catch (error) {
            dispatch(advertsLoadedFailure(error));
            throw error;
        }
    };

export const advertsLoadedRequest = () => ({
    type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = (adverts) => ({
    type: ADVERTS_LOADED_SUCCESS,
    payload: adverts,
});

export const advertsLoadedFailure = (error) => ({
    type: ADVERTS_LOADED_FAILURE,
    error: true,
    payload: error,
});

// ADVERT DETAIL thunk & actions:

export const advertLoadedRequest = () => ({
    type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = (adverts) => ({
    type: ADVERT_LOADED_SUCCESS,
    payload: adverts,
});

export const advertLoadedFailure = (error) => ({
    type: ADVERT_LOADED_FAILURE,
    error: true,
    payload: error,
});

export const advertLoad =
    (advertId) =>
    (dispatch, getState, { service: { advs } }) => {
        const isLoaded = getAdvertById(advertId)(getState());
        if (isLoaded) {
            return;
        }

        dispatch(advertsLoadedRequest());
        try {
            const advert = advs.getAdvert(advertId);
            dispatch(advertLoadedSuccess(advert));
        } catch (error) {
            dispatch(advertLoadedFailure(error));
        }
    };

//ADD TAGS thunk & actions:
export const getTagsListed =
    () =>
    async (dispatch, getState, { service: { advs } }) => {
        if (areTagsLoaded(getState())) {
            return;
        }
        dispatch(tagListRequest());
        try {
            const tags = await advs.getTagList();
            dispatch(tagListSuccess(tags));
        } catch (error) {
            dispatch(tagListFailure(error));
            throw error;
        }
    };

export const tagListRequest = () => ({
    type: ADD_TAGS_REQUEST,
});

export const tagListSuccess = (tags) => ({
    type: ADD_TAGS_SUCCESS,
    payload: tags,
});

export const tagListFailure = (error) => ({
    type: ADD_TAGS_FAILURE,
    payload: error,
    error: true,
});

//CREATE ADVERT thunk & actions:
export const advertCreatedRequest = () => ({
    type: ADVERT_CREATED_REQUEST,
});

export const advertCreatedSuccess = (advert) => ({
    type: ADVERT_CREATED_SUCCESS,
    payload: advert,
});

export const advertCreatedFailure = (error) => ({
    type: ADVERT_CREATED_FAILURE,
    error: true,
    payload: error,
});

export const advertCreate =
    (advert) =>
    async (dispatch, _getState, { service: { advs }, router }) => {
        dispatch(advertCreatedRequest());
        try {
            const createdAdvert = await advs.createNewAdvert(advert);
            dispatch(advertCreatedSuccess(createdAdvert));
            router.navigate(`/adverts/${createdAdvert.id}`);
            return createdAdvert;
        } catch (error) {
            advertCreatedFailure(error);
            throw error;
        }
    };

//DELETE ADVERT thunk & actions:

export const advertDeletedRequest = () => ({
    type: ADVERT_DELETED_REQUEST,
});

export const advertDeletedSuccess = (advertId) => ({
    type: ADVERT_DELETED_SUCCESS,
    payload: advertId,
});

export const advertDeletedFailure = (error) => ({
    type: ADVERT_DELETED_FAILURE,
    error: true,
    payload: error,
});

export const deletedAdvert =
    (advertId) =>
    async (dispatch, _getState, { service: { advs }, router }) => {
        dispatch(advertDeletedRequest());
        try {
            await advs.deleteAdvert(advertId);
            dispatch(advertDeletedSuccess(advertId));
            router.navigate('/');
        } catch (error) {
            dispatch(advertDeletedFailure(error));
        }
    };

export const adFilterName = (value) => ({
    type: ADV_FILTER_NAME,
    payload: value,
});

export const adFilterSale = (value) => ({
    type: ADV_FILTER_SALE,
    payload: value,
});

export const adFilterTags = (value) => ({
    type: ADV_FILTER_TAGS,
    payload: value,
});

/*export const adFilterMinPrice = (value) => ({
    type: ADV_FILTER_MIN_PRICE,
    payload: value,
});*/

/*export const adFilterMaxPrice = (value) => ({
    type: ADV_FILTER_MAX_PRICE,
    payload: value,
});*/

export const adFilterPrice = (minPrice, maxPrice, exactPrice) => ({
    type: ADV_FILTER_PRICE,
    payload: { minPrice, maxPrice },
});
export function toggleResult(value) {
    return {
        type: TOGGLE_RESULT,
        value, // Show result?
    };
}

//MODAL WINDOWS

export const openModal = (modalType) => ({
    type: OPEN_MODAL,
    payload: modalType
});

export const closeModal = () => ({
    type: CLOSE_MODAL,
});
