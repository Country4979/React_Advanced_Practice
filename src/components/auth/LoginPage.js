import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    authLogin,
    authLoginRequest,
    authlogout,
    closeModal,
    openModal,
} from '../../store/actions';
import Button from '../shared/Button';
import './LoginPage.css';
import '../shared/Buttons.css';
//import { UseModal } from '../modals/UseModal';
import Modal from '../modals/Modal';
import { getIsLogged, getModalstate, getUi } from '../../store/selectors';

const LoginPage = () => {
    const isLogged = useSelector(getIsLogged);
    const [credentials, setCredentials] = useState({
        email: '',
        password: '',
        rememberMe: false,
    });
    const dispatch = useDispatch();

    const { isLoading, error } = useSelector(getUi);
    const modalWindow = useSelector(getModalstate);
    const [errorMs, setErrorMs] = useState('');

    const onLogout = () => dispatch(authlogout());

    const handleSubmit = async (event) => {
        event.preventDefault();

        dispatch(authLoginRequest());
        try {
            await dispatch(authLogin(credentials));
        } catch (error) {
            error.message === 'Network Error'
                ? setErrorMs('An error occurred while logging in')
                : setErrorMs('Incorrect username or password');
            dispatch(openModal('error')); // open error modal window
        }
    };

    const closeModals = () => {
        dispatch(closeModal()); // close modal modal window
    };

    const handleChange = (event) => {
        const { name, value, type, checked } = event.target;
        if (type === 'checkbox') {
            setCredentials((prevCredentials) => ({
                ...prevCredentials,
                [name]: checked,
            }));
        } else {
            setCredentials((prevCredentials) => ({
                ...prevCredentials,
                [name]: value,
            }));
        }
    };

    const handleClickLogout = async () => {
        onLogout();
    };

    const buttonDisabled =
        isLoading || !credentials.email || !credentials.password;
    const logoutButtonDisabbled = isLogged;

    return (
        <>
            {modalWindow.isOpen && (
                <Modal
                    isOpen={modalWindow.isOpen}
                    closeModal={closeModals}
                    modalType={modalWindow.modalType}
                >
                    {modalWindow.modalType === 'success' ? (
                        <>
                            <h3 className='modalErrorH3'>Successful login!!</h3>
                            <small>You will be re-directed.</small>
                        </>
                    ) : (
                        <>
                            <h3 className='modalErrorH3'>
                                {error.message === 'Network Error'
                                    ? 'An error occurred while logging in'
                                    : 'Incorrect username or password'}
                                .
                            </h3>
                            <Button
                                className='noDeleteButton'
                                variant='primary'
                                onClick={closeModals}
                            >
                                Please try again...
                            </Button>
                        </>
                    )}
                </Modal>
            )}
            <div className='infoContainer'>
                <div className='leftSide' id='leftSide'>
                    <h1 id='textLogin'>Already Logged?</h1>
                    <p>
                        Please{' '}
                        <Button
                            data-testid='logoutButton'
                            className='noDeleteButton'
                            variant='primary'
                            onClick={handleClickLogout}
                            disabled={!logoutButtonDisabbled}
                        >
                            logout
                        </Button>{' '}
                        before to access AlaPop.
                    </p>
                </div>

                <div className='rigthSide'>
                    <form id='logUser' onSubmit={handleSubmit}>
                        <label htmlFor='email'>email:</label>
                        <input
                            data-testid='email'
                            type='email'
                            name='email'
                            onChange={handleChange}
                            value={credentials.email}
                        />
                        <br />
                        <label htmlFor='password'>Password:</label>
                        <input
                            data-testid='password'
                            type='password'
                            name='password'
                            onChange={handleChange}
                            value={credentials.password}
                        />
                        <br />
                        <Button
                            data-testid='loginButton'
                            className='noDeleteButton'
                            variant='primary'
                            type='submit'
                            disabled={buttonDisabled}
                        >
                            Login
                        </Button>
                        <br />
                        <label>
                            Remember Login?
                            <input
                                type='checkbox'
                                name='rememberMe'
                                checked={credentials.rememberMe}
                                onChange={handleChange}
                            />
                        </label>
                    </form>
                </div>
            </div>
        </>
    );
};

export default LoginPage;
