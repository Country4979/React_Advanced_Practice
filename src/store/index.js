import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';
import * as auth from '../components/auth/service';
import * as advs from '../components/adverts/service';
import * as client from '../api/client';
import * as reducers from './reducers';
import * as actionCreators from './actions';

const reducer = combineReducers(reducers);
const composeEnhancers = composeWithDevTools({
    actionCreators,
});
const failureRedirects = (router, redirectsMap) => () => (next) => (action) => {
    const result = next(action);
    if (action.error) {
        const redirect = redirectsMap[action.payload];
        if (redirect) {
            router.navigate(redirect);
        }
    }
    return result;
};

export default function configureStore(preloadedState, { router }) {
    const middleware = [
        thunk.withExtraArgument({ service: { auth, advs }, client, router }),
        failureRedirects(router, { 401: '/login', 404: '/404' }),
    ];
    const store = createStore(
        reducer,
        preloadedState,
        composeEnhancers(applyMiddleware(...middleware))
    );
    return store;
}
