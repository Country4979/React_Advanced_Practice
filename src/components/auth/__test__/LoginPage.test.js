import { fireEvent, render, screen } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers';
import { authLogin } from '../../../store/actions';

jest.mock('../../../store/actions');
jest.mock('../../modals/Modal', () => {
    //Creates a simulated function to check the properties passed to the Modal component
    const mockModal = jest.fn((props) => null);
    return mockModal;
});
describe('LoginPage', () => {
    const store = {
        getState: () => defaultState,
        subscribe: () => jest.fn(),
        dispatch: () => {},
    };
    const renderComponent = () =>
        render(
            <Provider store={store}>
                <LoginPage />
            </Provider>
        );

    test('snapshot', () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });

    test('should dispatch authLogin action, ()', () => {
        const email = 'test@test.com';
        const password = '123';
        let rememberMe = false;

        renderComponent();

        const emailInput = screen.getByTestId(/email/);
        const passwordInput = screen.getByTestId(/password/);
        const checkboxInput = screen.getByRole('checkbox');
        const submitLoginButton = screen.getByTestId(/loginButton/);
        const submitLogoutButton = screen.getByTestId(/logoutButton/);

        //Testing disabled buttons
        expect(submitLoginButton).toBeDisabled();
        expect(submitLogoutButton).toBeDisabled();

        //Events for inputs
        fireEvent.change(emailInput, { target: { value: email } });
        fireEvent.change(passwordInput, { target: { value: password } });

        //Testing Enabled/disabled buttons
        expect(submitLoginButton).toBeEnabled();
        expect(submitLogoutButton).toBeDisabled();

        //Testing Click on Button/Checkbox
        fireEvent.click(submitLoginButton);
        fireEvent.click(checkboxInput);
        fireEvent.change(checkboxInput, { target: { value: rememberMe } });

        expect(authLogin).toHaveBeenCalledWith({ email, password, rememberMe });
    });
});
