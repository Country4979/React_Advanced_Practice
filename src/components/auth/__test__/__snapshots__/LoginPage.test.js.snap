// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="infoContainer"
  >
    <div
      class="leftSide"
      id="leftSide"
    >
      <h1
        id="textLogin"
      >
        Already Logged?
      </h1>
      <p>
        Please
         
        <button
          class="sc-beqWaB dMDojR noDeleteButton"
          data-testid="logoutButton"
          disabled=""
        >
          logout
        </button>
         
        before to access AlaPop.
      </p>
    </div>
    <div
      class="rigthSide"
    >
      <form
        id="logUser"
      >
        <label
          for="email"
        >
          email:
        </label>
        <input
          data-testid="email"
          name="email"
          type="email"
          value=""
        />
        <br />
        <label
          for="password"
        >
          Password:
        </label>
        <input
          data-testid="password"
          name="password"
          type="password"
          value=""
        />
        <br />
        <button
          class="sc-beqWaB dMDojR noDeleteButton"
          data-testid="loginButton"
          disabled=""
          type="submit"
        >
          Login
        </button>
        <br />
        <label>
          Remember Login?
          <input
            name="rememberMe"
            type="checkbox"
          />
        </label>
      </form>
    </div>
  </div>
</div>
`;
